File: app/src/androidTest/java/com/poyka/ripdpi/e2e/TestSocketBinder.java
package com.poyka.ripdpi.e2e;

import android.os.Bundle;
import java.nio.charset.StandardCharsets;

final class TestSocketBinder {
    private static final int RESULT_OK = 0;
    private static final int RESULT_RESPONSE = 1;
    private static final int RESULT_BOUND_DEVICE = 2;
    private static final int RESULT_FAILURE_KIND = 3;
    private static final int RESULT_ERRNO = 4;
    private static final int RESULT_LOCAL_PORT = 5;
    private static final int RESULT_FAILURE_STAGE = 6;
    private static final int RESULT_WIDTH = 7;
    private static final String EXTRA_OK = "ok";
    private static final String EXTRA_RESPONSE = "response";
    private static final String EXTRA_BOUND_DEVICE = "bound_device";
    private static final String EXTRA_FAILURE_KIND = "failure_kind";
    private static final String EXTRA_ERRNO = "errno";
    private static final String EXTRA_LOCAL_PORT = "local_port";
    private static final String EXTRA_FAILURE_STAGE = "failure_stage";

    static {
        System.loadLibrary("test_socket_binder");
    }

    private TestSocketBinder() {}

    static void tcpRoundTrip(
            String host,
            int port,
            String payload,
            int connectTimeoutMs,
            int readTimeoutMs,
            String device,
            Bundle extras) {
        writeResult(
                extras,
                nativeTcpRoundTrip(
                        host,
                        port,
                        payload.getBytes(StandardCharsets.UTF_8),
                        connectTimeoutMs,
                        readTimeoutMs,
                        device.getBytes(StandardCharsets.UTF_8)));
    }

    static void udpRoundTrip(
            String host,
            int port,
            String payload,
            int timeoutMs,
            String device,
            Bundle extras) {
        writeResult(
                extras,
                nativeUdpRoundTrip(
                        host,
                        port,
                        payload.getBytes(StandardCharsets.UTF_8),
                        timeoutMs,
                        device.getBytes(StandardCharsets.UTF_8)));
    }

    private static void writeResult(Bundle extras, String[] result) {
        if (result == null || result.length != RESULT_WIDTH) {
            throw new IllegalStateException("Unexpected native socket probe result width");
        }
        extras.putBoolean(EXTRA_OK, Boolean.parseBoolean(result[RESULT_OK]));
        putString(extras, EXTRA_RESPONSE, result[RESULT_RESPONSE]);
        putString(extras, EXTRA_BOUND_DEVICE, result[RESULT_BOUND_DEVICE]);
        putString(extras, EXTRA_FAILURE_KIND, result[RESULT_FAILURE_KIND]);
        putInteger(extras, EXTRA_ERRNO, result[RESULT_ERRNO]);
        putInteger(extras, EXTRA_LOCAL_PORT, result[RESULT_LOCAL_PORT]);
        putString(extras, EXTRA_FAILURE_STAGE, result[RESULT_FAILURE_STAGE]);
    }

    private static void putString(Bundle extras, String key, String value) {
        if (value != null) {
            extras.putString(key, value);
        }
    }

    private static void putInteger(Bundle extras, String key, String value) {
        if (value != null) {
            extras.putInt(key, Integer.parseInt(value));
        }
    }

    private static native String[] nativeTcpRoundTrip(
            String host,
            int port,
            byte[] payload,
            int connectTimeoutMs,
            int readTimeoutMs,
            byte[] deviceUtf8);

    private static native String[] nativeUdpRoundTrip(
            String host, int port, byte[] payload, int timeoutMs, byte[] deviceUtf8);
}

File: app/src/androidTest/kotlin/com/poyka/ripdpi/e2e/TestSocketBinder.kt
package com.poyka.ripdpi.e2e

import android.os.Bundle

internal object TestSocketBinder {
    private const val ResultOk = 0
    private const val ResultResponse = 1
    private const val ResultBoundDevice = 2
    private const val ResultFailureKind = 3
    private const val ResultErrno = 4
    private const val ResultLocalPort = 5
    private const val ResultFailureStage = 6
    private const val ResultWidth = 7

    private const val ExtraOk = "ok"
    private const val ExtraResponse = "response"
    private const val ExtraBoundDevice = "bound_device"
    private const val ExtraFailureKind = "failure_kind"
    private const val ExtraErrno = "errno"
    private const val ExtraLocalPort = "local_port"
    private const val ExtraFailureStage = "failure_stage"

    init {
        System.loadLibrary("test_socket_binder")
    }

    fun tcpRoundTrip(
        host: String,
        port: Int,
        payload: String,
        connectTimeoutMs: Int,
        readTimeoutMs: Int,
        device: String,
        extras: Bundle,
    ) = writeResult(
        extras,
        nativeTcpRoundTrip(
            host,
            port,
            payload.encodeToByteArray(),
            connectTimeoutMs,
            readTimeoutMs,
            device.encodeToByteArray(),
        ),
    )

    fun udpRoundTrip(
        host: String,
        port: Int,
        payload: String,
        timeoutMs: Int,
        device: String,
        extras: Bundle,
    ) = writeResult(
        extras,
        nativeUdpRoundTrip(host, port, payload.encodeToByteArray(), timeoutMs, device.encodeToByteArray()),
    )

    private fun writeResult(
        extras: Bundle,
        result: Array<String>?,
    ) {
        check(result?.size == ResultWidth) { "Unexpected native socket probe result width" }
        extras.putBoolean(ExtraOk, result[ResultOk].toBoolean())
        extras.putStringOrSkip(ExtraResponse, result[ResultResponse])
        extras.putStringOrSkip(ExtraBoundDevice, result[ResultBoundDevice])
        extras.putStringOrSkip(ExtraFailureKind, result[ResultFailureKind])
        extras.putIntOrSkip(ExtraErrno, result[ResultErrno])
        extras.putIntOrSkip(ExtraLocalPort, result[ResultLocalPort])
        extras.putStringOrSkip(ExtraFailureStage, result[ResultFailureStage])
    }

    private fun Bundle.putStringOrSkip(
        key: String,
        value: String?,
    ) {
        value?.let { putString(key, it) }
    }

    private fun Bundle.putIntOrSkip(
        key: String,
        value: String?,
    ) {
        value?.toInt()?.let { putInt(key, it) }
    }

    @JvmStatic
    private external fun nativeTcpRoundTrip(
        host: String,
        port: Int,
        payload: ByteArray,
        connectTimeoutMs: Int,
        readTimeoutMs: Int,
        deviceUtf8: ByteArray,
    ): Array<String>?

    @JvmStatic
    private external fun nativeUdpRoundTrip(
        host: String,
        port: Int,
        payload: ByteArray,
        timeoutMs: Int,
        deviceUtf8: ByteArray,
    ): Array<String>?
}
